source/extensions/network/dns_resolver/getaddrinfo/getaddrinfo.cc

```diff
@@ -0,0 +1,217 @@
+
+  ~GetAddrInfoDnsResolver() override {
+    {
+      absl::MutexLock guard(&mutex_);
+      shutting_down_ = true;
+    }
+
+    resolver_thread_->join();
+  }
+
+  // DnsResolver
+  ActiveDnsQuery* resolve(const std::string& dns_name, DnsLookupFamily dns_lookup_family,
+                          ResolveCb callback) override {
+    ENVOY_LOG(debug, "adding new query [{}] to pending queries", dns_name);
+    auto new_query = std::make_unique<PendingQuery>(dns_name, dns_lookup_family, callback);
+    absl::MutexLock guard(&mutex_);
+    pending_queries_.emplace_back(std::move(new_query));
+    return pending_queries_.back().get();
+  }
+
+  void resetNetworking() override {}
+
+private:
+  class PendingQuery : public ActiveDnsQuery {
+  public:
+    PendingQuery(const std::string& dns_name, DnsLookupFamily dns_lookup_family, ResolveCb callback)
+        : dns_name_(dns_name), dns_lookup_family_(dns_lookup_family), callback_(callback) {}
+
+    void cancel(CancelReason) override {
+      ENVOY_LOG(debug, "cancelling query [{}]", dns_name_);
+      cancelled_ = true;
+    }
+
+    const std::string dns_name_;
+    const DnsLookupFamily dns_lookup_family_;
+    ResolveCb callback_;
+    bool cancelled_{false};
+  };
+  // Must be a shared_ptr for passing around via post.
+  using PendingQuerySharedPtr = std::shared_ptr<PendingQuery>;
+
+  // Parse a GAI response and determine the final address list.
+  std::pair<ResolutionStatus, std::list<DnsResponse>>
+  processGaiResponse(const PendingQuery& query, addrinfo* addrinfo_result) {
+    std::list<DnsResponse> v4_results;
+    std::list<DnsResponse> v6_results;
+    for (auto ai = addrinfo_result; ai != nullptr; ai = ai->ai_next) {
+      if (ai->ai_family == AF_INET) {
+        sockaddr_in address;
+        memset(&address, 0, sizeof(address));
+        address.sin_family = AF_INET;
+        address.sin_port = 0;
+        address.sin_addr = reinterpret_cast<sockaddr_in*>(ai->ai_addr)->sin_addr;
+
+        v4_results.emplace_back(DnsResponse(std::make_shared<const Address::Ipv4Instance>(&address),
+                                            std::chrono::seconds(60)));
+      } else if (ai->ai_family == AF_INET6) {
+        sockaddr_in6 address;
+        memset(&address, 0, sizeof(address));
+        address.sin6_family = AF_INET6;
+        address.sin6_port = 0;
+        address.sin6_addr = reinterpret_cast<sockaddr_in6*>(ai->ai_addr)->sin6_addr;
+        v6_results.emplace_back(DnsResponse(std::make_shared<const Address::Ipv6Instance>(address),
+                                            std::chrono::seconds(60)));
+      }
+    }
+    Api::OsSysCallsSingleton::get().freeaddrinfo(addrinfo_result);
+
+    std::list<DnsResponse> final_results;
+    switch (query.dns_lookup_family_) {
+    case DnsLookupFamily::All: {
+      final_results = std::move(v4_results);
+      final_results.splice(final_results.begin(), v6_results);
```

Comment:
defering to @RyanTheOptimist re happy eyeballs sorting

Reply:
The sorting is done at a high layer so I don't think it matters for this.